File: cpp/src/tsp.cpp
#include "tsp.hpp"

#include <algorithm>
#include <cmath>
#include <numeric>
#include <set>

namespace antchain {

double TSPInstance::tour_length(const std::vector<int>& tour) const {
    double total = 0.0;
    const int n = static_cast<int>(tour.size());
    for (int i = 0; i < n; ++i) {
        total += d(tour[i], tour[(i + 1) % n]);
    }
    return total;
}

bool TSPInstance::is_valid_tour(const std::vector<int>& tour) const {
    if (static_cast<int>(tour.size()) != n_cities) return false;
    std::vector<bool> seen(n_cities, false);
    for (int c : tour) {
        if (c < 0 || c >= n_cities || seen[c]) return false;
        seen[c] = true;
    }
    return true;
}

TSPInstance generate_instance(uint64_t seed, int n_cities) {
    SplitMix64 rng(seed);
    TSPInstance inst;
    inst.n_cities = n_cities;
    inst.xs.resize(n_cities);
    inst.ys.resize(n_cities);
    for (int i = 0; i < n_cities; ++i) {
        inst.xs[i] = rng.next_double();
        inst.ys[i] = rng.next_double();
    }
    inst.dist.assign(static_cast<size_t>(n_cities) * n_cities, 0.0);
    for (int i = 0; i < n_cities; ++i) {
        for (int j = 0; j < n_cities; ++j) {
            double dx = inst.xs[i] - inst.xs[j];
            double dy = inst.ys[i] - inst.ys[j];
            inst.dist[static_cast<size_t>(i) * n_cities + j] = std::sqrt(dx * dx + dy * dy);
        }
    }
    return inst;
}

std::vector<int> nearest_neighbor_tour(const TSPInstance& inst, int start) {
    const int n = inst.n_cities;
    std::vector<bool> visited(n, false);
    std::vector<int> tour;
    tour.reserve(n);
    tour.push_back(start);
    visited[start] = true;
    int current = start;
    for (int step = 1; step < n; ++step) {
        int best = -1;
        double best_d = 1e18;
        for (int c = 0; c < n; ++c) {
            if (!visited[c] && inst.d(current, c) < best_d) {
                best_d = inst.d(current, c);
                best = c;
            }
        }
        tour.push_back(best);
        visited[best] = true;
        current = best;
    }
    return tour;
}

AntColonyOptimizer::AntColonyOptimizer(const TSPInstance& inst, uint64_t seed, int n_ants,
                                        double alpha, double beta, double rho, double q)
    : inst_(&inst), n_ants_(n_ants), alpha_(alpha), beta_(beta), rho_(rho), q_(q), rng_(seed) {
    reset(inst);
}

void AntColonyOptimizer::reset(const TSPInstance& inst) {
    inst_ = &inst;
    const int n = inst.n_cities;
    tau_.assign(static_cast<size_t>(n) * n, 1.0);
    eta_.assign(static_cast<size_t>(n) * n, 0.0);
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            double dij = inst.d(i, j);
            eta_[static_cast<size_t>(i) * n + j] = dij > 0.0 ? 1.0 / dij : 0.0;
        }
    }
    best_tour.clear();
    best_length = 1e18;
}

void AntColonyOptimizer::rebind_instance_keep_pheromones(const TSPInstance& inst) {
    inst_ = &inst;
    const int n = inst.n_cities;
    eta_.assign(static_cast<size_t>(n) * n, 0.0);
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            double dij = inst.d(i, j);
            eta_[static_cast<size_t>(i) * n + j] = dij > 0.0 ? 1.0 / dij : 0.0;
        }
    }
    best_tour.clear();
    best_length = 1e18;
}

std::vector<int> AntColonyOptimizer::construct_tour() {
    const int n = inst_->n_cities;
    std::vector<bool> visited(n, false);
    std::vector<int> tour;
    tour.reserve(n);
    int start = rng_.next_int(n);
    tour.push_back(start);
    visited[start] = true;
    int current = start;

    std::vector<double> weights(n);
    for (int step = 1; step < n; ++step) {
        double total = 0.0;
        for (int c = 0; c < n; ++c) {
            if (visited[c]) {
                weights[c] = 0.0;
                continue;
            }
            double tau = tau_[static_cast<size_t>(current) * n + c];
            double eta = eta_[static_cast<size_t>(current) * n + c];
            weights[c] = std::pow(tau, alpha_) * std::pow(eta, beta_);
            total += weights[c];
        }
        int next = -1;
        if (total <= 0.0) {
            // no distinguishing weight (degenerate case) -- pick uniformly
            // among unvisited cities
            int remaining = n - step;
            int pick = rng_.next_int(remaining);
            for (int c = 0; c < n; ++c) {
                if (!visited[c]) {
                    if (pick == 0) { next = c; break; }
                    --pick;
                }
            }
        } else {
            double r = rng_.next_double() * total;
            double cum = 0.0;
            for (int c = 0; c < n; ++c) {
                if (visited[c]) continue;
                cum += weights[c];
                if (r <= cum) { next = c; break; }
            }
            if (next == -1) {
                for (int c = n - 1; c >= 0; --c) if (!visited[c]) { next = c; break; }
            }
        }
        tour.push_back(next);
        visited[next] = true;
        current = next;
    }
    return tour;
}

std::pair<std::vector<int>, double> AntColonyOptimizer::step(int n_iterations) {
    const int n = inst_->n_cities;
    for (int it = 0; it < n_iterations; ++it) {
        std::vector<std::vector<int>> tours(n_ants_);
        std::vector<double> lengths(n_ants_);
        for (int k = 0; k < n_ants_; ++k) {
            tours[k] = construct_tour();
            lengths[k] = inst_->tour_length(tours[k]);
        }
        for (auto& v : tau_) v *= (1.0 - rho_);
        for (int k = 0; k < n_ants_; ++k) {
            if (lengths[k] <= 0.0) continue;
            double deposit = q_ / lengths[k];
            const auto& t = tours[k];
            for (int i = 0; i < n; ++i) {
                int a = t[i];
                int b = t[(i + 1) % n];
                tau_[static_cast<size_t>(a) * n + b] += deposit;
                tau_[static_cast<size_t>(b) * n + a] += deposit;
            }
        }
        int best_idx = static_cast<int>(std::min_element(lengths.begin(), lengths.end()) - lengths.begin());
        if (lengths[best_idx] < best_length) {
            best_length = lengths[best_idx];
            best_tour = tours[best_idx];
        }
    }
    return {best_tour, best_length};
}

} // namespace antchain

File: sim/antchain_sim/tsp.py
"""TSP instance generation, tour evaluation, and an Ant Colony Optimization solver.

Instances are generated deterministically from a seed (in AntChain, the seed is
derived from the previous block hash) so that every miner sees the same
instance at the same time with no precomputation advantage.
"""
from __future__ import annotations

import hashlib
from dataclasses import dataclass, field

import numpy as np


def seed_from_bytes(data: bytes) -> int:
    """Derive a 63-bit integer seed from arbitrary bytes (e.g. a block hash)."""
    digest = hashlib.sha256(data).digest()
    return int.from_bytes(digest[:8], "big") & 0x7FFFFFFFFFFFFFFF


@dataclass
class TSPInstance:
    n_cities: int
    coords: np.ndarray  # (n_cities, 2)
    dist: np.ndarray = field(repr=False)  # (n_cities, n_cities)
    instance_id: str

    def tour_length(self, tour: np.ndarray) -> float:
        """O(n) cost of a candidate tour (array of city indices, a permutation)."""
        return float(self.dist[tour, np.roll(tour, -1)].sum())

    def is_valid_tour(self, tour: np.ndarray) -> bool:
        return (
            len(tour) == self.n_cities
            and len(set(tour.tolist())) == self.n_cities
            and tour.min() == 0
            and tour.max() == self.n_cities - 1
        )


def generate_instance(seed_bytes: bytes, n_cities: int = 30) -> TSPInstance:
    """Deterministic instance generator Gen(H(block)). Public and reproducible."""
    seed = seed_from_bytes(seed_bytes)
    rng = np.random.default_rng(seed)
    coords = rng.random((n_cities, 2))
    diff = coords[:, None, :] - coords[None, :, :]
    dist = np.sqrt((diff ** 2).sum(axis=-1))
    instance_id = hashlib.sha256(seed_bytes).hexdigest()[:16]
    return TSPInstance(n_cities=n_cities, coords=coords, dist=dist, instance_id=instance_id)


def nearest_neighbor_tour(instance: TSPInstance, start: int = 0) -> np.ndarray:
    """Cheap reference heuristic used to initialize f_old / T_opt."""
    n = instance.n_cities
    unvisited = set(range(n))
    tour = [start]
    unvisited.remove(start)
    current = start
    while unvisited:
        nxt = min(unvisited, key=lambda c: instance.dist[current, c])
        tour.append(nxt)
        unvisited.remove(nxt)
        current = nxt
    return np.array(tour)


def two_opt(instance: TSPInstance, tour: np.ndarray, max_passes: int = 5) -> np.ndarray:
    """Local-search refinement used only to estimate a strong reference optimum
    for the convergence-quality metric (not part of the mining protocol)."""
    tour = tour.copy()
    n = len(tour)
    dist = instance.dist
    for _ in range(max_passes):
        improved = False
        for i in range(n - 1):
            for j in range(i + 2, n):
                if i == 0 and j == n - 1:
                    continue
                a, b = tour[i], tour[i + 1]
                c, d = tour[j], tour[(j + 1) % n]
                delta = (dist[a, c] + dist[b, d]) - (dist[a, b] + dist[c, d])
                if delta < -1e-12:
                    tour[i + 1:j + 1] = tour[i + 1:j + 1][::-1]
                    improved = True
        if not improved:
            break
    return tour


def estimate_reference_optimum(instance: TSPInstance, n_starts: int = 6) -> float:
    """Best-of-several nearest-neighbor + 2-opt runs; used only for the
    convergence-quality metric, never by the protocol itself."""
    best = np.inf
    n = instance.n_cities
    starts = np.linspace(0, n - 1, num=min(n_starts, n), dtype=int)
    for s in starts:
        t = nearest_neighbor_tour(instance, start=int(s))
        t = two_opt(instance, t)
        length = instance.tour_length(t)
        if length < best:
            best = length
    return float(best)


class AntColonyOptimizer:
    """Incremental ACO solver: run() can be called repeatedly with a small
    iteration budget so a miner's search can be interleaved with hash attempts
    in the discrete-time simulation loop.

    Pheromone state (`tau`) is the mechanism's memory. Protocol-honest miners
    call reset() at the start of every block round; a "pheromone-hoarder"
    strategic miner (see miner.py) deliberately skips reset() to carry an
    advantage across blocks/instances.
    """

    def __init__(self, instance: TSPInstance, n_ants: int = 12, alpha: float = 1.0,
                 beta: float = 3.0, rho: float = 0.1, q: float = 1.0, rng: np.random.Generator | None = None):
        self.instance = instance
        self.n_ants = n_ants
        self.alpha = alpha
        self.beta = beta
        self.rho = rho
        self.q = q
        self.rng = rng if rng is not None else np.random.default_rng()
        n = instance.n_cities
        self.tau = np.full((n, n), 1.0)
        with np.errstate(divide="ignore"):
            self.eta = np.where(instance.dist > 0, 1.0 / instance.dist, 0.0)
        self.best_tour: np.ndarray | None = None
        self.best_length: float = np.inf

    def reset(self, instance: TSPInstance | None = None) -> None:
        if instance is not None:
            self.instance = instance
            with np.errstate(divide="ignore"):
                self.eta = np.where(instance.dist > 0, 1.0 / instance.dist, 0.0)
        n = self.instance.n_cities
        self.tau = np.full((n, n), 1.0)
        self.best_tour = None
        self.best_length = np.inf

    def _construct_tour(self) -> np.ndarray:
        n = self.instance.n_cities
        start = int(self.rng.integers(n))
        visited = np.zeros(n, dtype=bool)
        visited[start] = True
        tour = np.empty(n, dtype=int)
        tour[0] = start
        current = start
        for step in range(1, n):
            mask = ~visited
            weights = (self.tau[current] ** self.alpha) * (self.eta[current] ** self.beta)
            weights = weights * mask
            total = weights.sum()
            if total <= 0:
                candidates = np.flatnonzero(mask)
                nxt = int(candidates[int(self.rng.integers(len(candidates)))])
            else:
                # roulette-wheel selection via a single uniform draw + cumsum
                # search, avoiding np.random.Generator.choice's per-call
                # overhead (probability-array validation, etc.) which
                # dominates runtime at the iteration counts this simulation needs.
                r = self.rng.random() * total
                cum = np.cumsum(weights)
                nxt = int(np.searchsorted(cum, r, side="right"))
                if nxt >= n:
                    nxt = int(np.flatnonzero(mask)[-1])
            tour[step] = nxt
            visited[nxt] = True
            current = nxt
        return tour

    def step(self, n_iterations: int = 1) -> tuple[np.ndarray, float]:
        """Run `n_iterations` ACO iterations (construct + update pheromones).
        Returns the best (tour, length) found so far across all calls since
        the last reset()."""
        n = self.instance.n_cities
        for _ in range(n_iterations):
            tours = [self._construct_tour() for _ in range(self.n_ants)]
            lengths = [self.instance.tour_length(t) for t in tours]
            self.tau *= (1 - self.rho)
            for t, length in zip(tours, lengths):
                if length <= 0:
                    continue
                deposit = self.q / length
                self.tau[t, np.roll(t, -1)] += deposit
                self.tau[np.roll(t, -1), t] += deposit
            best_idx = int(np.argmin(lengths))
            if lengths[best_idx] < self.best_length:
                self.best_length = lengths[best_idx]
                self.best_tour = tours[best_idx]
        return self.best_tour, self.best_length
